### PR TITLE
Intern BLS lazy public keys

### PR DESCRIPTION
**Pull request by hrxi**
This ensures that for each compressed public key there's at most one lazy public key, which means everyone profits when it is uncompressed.

This drops the (unused: never read from) BLS public key cache.

### CODE REVIEW

File: utils/src/interner.rs [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+    borrow::Borrow,
+    collections::HashSet,
+    hash::{Hash, Hasher},
+    ops::Deref,
+    sync::{Arc, Weak},
+};
+
+use parking_lot::RwLock;
+
+// TODO: document that you need to call notify_deletion in the `Drop` impl
```
Comment (paberr) [inline utils/src/interner.rs]: Document what you promised

**Review by paberr — COMMENTED** [review]
Think about whether we can drop the requirement of the user of the interner to implement a specific Drop by wrapping the type.
Also fix a couple of Clippy warnings.

File: utils/src/interner.rs (hunk 1 above)
Comment (hrxi) [reply utils/src/interner.rs]: No longer necessary.

**Comment by hrxi** [comment]
> Think about whether we can drop the requirement of the user of the interner to implement a specific Drop by wrapping the type.

Did so, resulted in a much nicer API. :)

File: bls/src/lazy.rs [hunk 2]
```diff
@@ -1,141 +1,110 @@
...
-    pub(crate) compressed: CompressedPublicKey,
-    pub(crate) cache: RwLock<Option<PublicKey>>,
+fn cache() -> &'static Interner<CompressedPublicKey, OnceLock<Option<PublicKey>>> {
+    static CACHE: OnceLock<Interner<CompressedPublicKey, OnceLock<Option<PublicKey>>>> =
+        OnceLock::new();
+    CACHE.get_or_init(Default::default)
 }
 
+#[derive(Clone)]
+pub struct LazyPublicKey(Interned<CompressedPublicKey, OnceLock<Option<PublicKey>>>);
```
Comment (jsdanielh) [inline bls/src/lazy.rs:20]: Could you add `rustdoc`?

File: lib/src/client.rs [hunk 3]
```diff
@@ -377,9 +376,7 @@ impl ClientInner {
             config.database,
         )?;
 
-        let bls_cache = Arc::new(Mutex::new(PublicKeyCache::new(
-            Policy::BLS_CACHE_MAX_CAPACITY,
-        )));
+        let bls_cache = Arc::new(Mutex::new(BlsCache::default()));
```
Comment (jsdanielh) [inline lib/src/client.rs:379]:
```suggestion
        let bls_cache = Arc::new(Mutex::new(BlsCache));
```

File: consensus/src/bls_cache.rs [hunk 4]
```diff
@@ -0,0 +1,25 @@
+use nimiq_bls::LazyPublicKey as BlsLazyPublicKey;
+use nimiq_primitives::policy::Policy;
+
+// TODO: implement some caching strategy
+pub struct BlsCache;
```
Comment (jsdanielh) [inline consensus/src/bls_cache.rs]: Could you add some `rustdoc`?

File: consensus/src/bls_cache.rs [hunk 5]
```diff
@@ -0,0 +1,30 @@
+use hashlink::LruCache;
+use nimiq_bls::LazyPublicKey as BlsLazyPublicKey;
+use nimiq_primitives::policy::Policy;
+
+pub struct BlsCache {
```
Comment (jsdanielh) [inline consensus/src/bls_cache.rs:6]: Could you add some rustdoc?

**Comment by jsdanielh** [comment]
Also it looks like there are untouched filed with clippy warnings

**Comment by viquezclaudio** [comment]
With this changes, in Safari we get consensus in ~12secs

Previously, in Safari, we used to get consensus in ~23secs

File: lib/src/client.rs (hunk 3 above)
Comment (hrxi) [reply lib/src/client.rs:379]: Was intentionally not done because I wanted to add some `struct` members later.

File: bls/src/lazy.rs (hunk 2 above)
Comment (hrxi) [reply bls/src/lazy.rs:20]: Done.

File: consensus/src/bls_cache.rs (hunk 5 above)
Comment (hrxi) [reply consensus/src/bls_cache.rs:6]: Done.

**Comment by hrxi** [comment]
> Also it looks like there are untouched filed with clippy warnings

Fixed.

**Review by jsdanielh — APPROVED** [review]
LGTM

**Pull request merged**